internal/services/ingest.go

```diff
@@ -228,344 +150,217 @@ func (m *ingestService) Run(ctx context.Context, startLedger uint32, endLedger u
 			err = fmt.Errorf("releasing advisory lock: %w", err)
 			log.Ctx(ctx).Error(err)
 		}
-		m.pool.Stop()
 	}()
 
-	// get latest ledger synced, to use as a cursor
-	var latestTrustlinesLedger uint32
-	if startLedger == 0 {
-		var err error
-		startLedger, err = m.models.IngestStore.GetLatestLedgerSynced(ctx, m.ledgerCursorName)
+	// Check if account tokens cache is populated
+	latestIngestedLedger, err := m.models.IngestStore.Get(ctx, m.ledgerCursorName)
+	if err != nil {
+		return fmt.Errorf("getting latest account-tokens ledger cursor: %w", err)
+	}
+
+	// If latestIngestedLedger == 0, then its an empty db. In that case, we get the latest checkpoint ledger
+	// and start from there.
+	if latestIngestedLedger == 0 {
+		startLedger, err = m.calculateCheckpointLedger(startLedger)
 		if err != nil {
-			return fmt.Errorf("getting latest ledger synced: %w", err)
+			return fmt.Errorf("calculating checkpoint ledger: %w", err)
 		}
 
-		latestTrustlinesLedger, err = m.models.IngestStore.GetLatestLedgerSynced(ctx, m.accountTokensCursorName)
-		if err != nil {
-			return fmt.Errorf("getting latest account-tokens ledger cursor synced: %w", err)
+		log.Ctx(ctx).Infof("Account tokens cache not populated, using checkpoint ledger: %d", startLedger)
+
+		if populateErr := m.accountTokenService.PopulateAccountTokens(ctx, startLedger); populateErr != nil {
+			return fmt.Errorf("populating account tokens cache: %w", populateErr)
+		}
+	} else {
+		// If we already have data ingested currently, then we check the start ledger value supplied by the user.
+		// If it is 0 or beyond the current ingested ledger, we just start from where we left off.
+		if startLedger == 0 || startLedger >= latestIngestedLedger {
+			startLedger = latestIngestedLedger + 1
+		} else {
+			// If start ledger is some value less than latest ingested ledger, we go into backfilling mode. In this mode
+			// we dont update the account token cache (since it is already populated with recent checkpoint ledger) and we dont
+			// update the latest ledger ingested cursor.
+			// NOTE: Currently we dont have the functionality of detecting gaps and intelligently backfilling so we would process the same
+			// ledgers again during backfilling. However the db insertions have ON CONFLICT DO NOTHING, so we would not do repeated insertions
+			m.backfillMode = true
 		}
 	}
 
 	signalChan := make(chan os.Signal, 1)
```

Comment:
I don't think it is appropriate to set up the signals here. You probably want to do this at the very top level when running the application. The context which is passed into Run() will be sufficient as a termination singal. You can use https://pkg.go.dev/os/signal#NotifyContext to create a context which is done once it receives SIGTERM, SIGINT, etc